dev/core#5734 create a FileCache implementation

herbdool opened the pull request:
Fixes https://lab.civicrm.org/dev/core/-/issues/5734

This probably needs some more work in terms of making it more efficient, but it's getting there. It works on my local copy.

herbdool committed:
dev/core#5734 create a FileCache implementation

Sjord commented:
The `get` method seems to have some logic to detect and handle the situation where `set` is still busy writing. I think it would be better if `set` does atomic writing, by first creating a temporary file and then renaming that to the correct file name.

Does `getCacheFile` need to guard against path traversal?

herbdool commented:
@Sjord thanks for reviewing.

> The get method seems to have some logic to detect and handle the situation where set is still busy writing. I think it would be better if set does atomic writing, by first creating a temporary file and then renaming that to the correct file name.

Actually, I'm not sure how important that logic is. I had been inspired by the Drupal 7 version of https://www.drupal.org/project/filecache, which included that logic. But I notice that the Drupal 8-11+ version doesn't have it at all. It just has `file_get_contents()` and returns NULL if it not successful. So I could also just remove it and we could add in atomic writing or some other method if it becomes an issue.

```
  protected function getFile(string $filename): ?\stdClass {
    if (is_file($filename)) {
      $serialized_contents = file_get_contents($filename);
      if ($serialized_contents !== FALSE) {
        $item = $this->serializer->decode($serialized_contents);
        // Only return the item if it could be successfully deserialized.
        if ($item !== FALSE) {
          return $item;
        }
      }
    }
    return NULL;
  }
```

> Does getCacheFile need to guard against path traversal?

I'm not sure. It's not taking in any user-supplied inputs; so is it a concern? I had taken a similar approach to `CRM_Extension_ClassLoader::getCacheFile()` which also doesn't appear to guard against path traversal.

totten commented:
> `ab -n 100 -c 10 'http://localhost/civicrm/event/info?id=1'`
>
> cache in db: 458 ms
> cache in file: 331 ms

Wow.

> Actually, I'm not sure how important that logic is. I had been inspired by the Drupal 7 version of https://www.drupal.org/project/filecache, which included that logic. But I notice that the Drupal 8-11+ version doesn't have it at a

Interesting. Their earlier implementation sounds better on paper... wonder why they dropped it. (Maybe they're counting on some lower-level behavior that makes it moot? Or maybe they have some higher-level locks that make it moot?)

In any case, there are mitigating factors:

* This implementation uses separate file-per-key (rather than file-per-group), so any contention would only arise on per-key basis.
* The importance of locking can vary by use-case.
* We can flag the driver as "Experimental".

My vote is - use whatever works well in your testing - but flag it "Experimental".

>  > Does getCacheFile need to guard against path traversal?
>
> I'm not sure. It's not taking in any user-supplied inputs; so is it a concern? I had taken a similar approach to `CRM_Extension_ClassLoader::getCacheFile()` which also doesn't appear to guard against path traversal.

Well, `CRM_Extension_ClassLoader::getCacheFile()` pushes all the dynamic data through `md5()`. So that's the main guard.

In this context, there is some regulation of `$key` when `set()`/`get()` call out to `CRM_Utils_Cache::assertValidKey()`, which [excludes](https://github.com/civicrm/civicrm-core/blob/6d737103d5ff98295f34b584be80f1933f22f952/CRM/Utils/Cache.php#L306) `/` and `\` (for [PSR-16](https://www.php-fig.org/psr/psr-16/#12-definitions) compliance),  ___but___ `assertValidKey()` usually isn't enforced in production. (This means the _de facto_ limits on key are driver-dependent.) I think something a little stronger is in order, e.g.

* Unconditional prohibition on scary chars
* `CRM_Utils_String::munge($key)`
* `md5($key)`

herbdool commented:
@totten 

What do you mean by 'flag the driver as "Experimental"'? I've added it to the settings template as "To use the filesystem (experimental) use a value of 'FileCache'". Is that enough?

I've added a couple methods to clean the key and clean the prefix. The key gets cleaned to replace any non-space, non-word character with underscore. And it also ensures that if the key is larger than 200, that it does an md5 of the remainder to ensure it stays within the limit. (I might have to doublecheck the logic since I borrowed it). And the prefix is cleaned too. The prefix included `/`s which was creating nested directories, but we only want it to be one directory for simplicity. So instead of `/angular/5.82.0/angularModules.txt` it's now `/angular_5_82_0/angularModules.txt`.

herbdool commented:
@colemanw I'm fine with not flagging it as experimental. I think it's working well enough to be used in production. I'm going to test it out on a couple sites of ours.

I recorded a few more samples using `ab -n 100 -c 10 'http://localhost/civicrm/event/info?id=1'`:

- db (ms): 411, 391, 413, 406
- file (ms): 370, 354, 361, 343

Average for db: 405 ms; average for file: 357 ms. So about 12% improvement.

totten commented:
> What is the end-user supposed to do when they see your feature is experimental? Use it? Not use it?

That's up to them! They're not babies! The word "Experimental" is in the dictionary. If the thing genuinely has unmanaged risks, then it's just honest communication. It invites feedback and rightfully scares the timid consumer away from that option.

Of course, context matters. Putting a big red "Experimental" flag on the payment UI would be much more onerous than some opt-in that's buried within `civicrm.settings.php`. And it's true that it's easy to become complacent (e.g. letting the low-key "experimental" flag linger even after years of widespread use).

But this too abstract.

--

The present issue is that, apparently, ___we don't understand file-locking___. Taking Drupal's FileCache as an example, we have mixed signals about what to do, with no indication of why the signals are mixed, and no real conclusion about what makes sense in our context.

I have definitely been bit before by this exact kind of thing (*ignoring file-locks -- finding that everything seems to work in testing -- and then two months later, realizing that it created some obscure/intermittent/hard-to-reproduce bug*). It's a very real risk, especially given that the caches serve several subsystems/use-cases.

It's also a normal (and arguably necessary) part of the process. Finding an upfront/analytic answer is expensive. There's nothing wrong with admitting that we don't know which locking strategy to use here. If the strategy is "ship it and see", that's cool. We just need to say so -- so that future bug-hunters are primed to recognize the issue.

herbdool commented:
> Their earlier implementation sounds better on paper... wonder why they dropped it. (Maybe they're counting on some lower-level behavior that makes it moot? Or maybe they have some higher-level locks that make it moot?)

@totten I couldn't find anything in the D8+ version that relies on lower-level or higher-level behavior. In the end, it's just using `file_get_content()` and `file_put_content()`. There are about 3-400 installs for the last four years or so, and I didn't see any bug reports about file locking or lack thereof. I didn't see any notes on why they decided to drop the file locking.

I was wondering about the risk with no file locking, or checking a file while it's being updated. If it can't read the file or can't unserialize then it returns the default value, and will need to build the result from scratch. As for writing, I read that the last PHP connection to call `file_put_content()` will be the one to write to the file, which seems reasonable. Most likely it would be the same data, or slightly newer.

herbdool commented:
I've been doing some more research. Turns out that Laravel's file caching includes file locking for both reading and writing, similar to D7 version. So I think I will incorporate it after all. File caching was the default for Laravel until 11.x so they've got lots of experience I think.

herbdool commented:
I've updated the PR. The update includes:

* using the `LOCK_EX` flag when using `file_put_contents()` - similar to Laravel's approach.
* added `getContents()` method which  replicates Laravel's and Drupal 7's filecache approach to getting the file content.
* the "in-memory cache" array now uses the prefix + key as the array key.
* use `\CRM_Utils_File::cleanDir()` to remove the "prefix" directory instead of handling file removal in the class.
* display error if `[civicrm.private]/filecache` cannot be created.

Comparing benchmarks is closer now, so the latest changes may have slowed down FileCache slightly. But I suppose it's a good tradeoff for more robustness, borrowing code from well-used implementations.

Sjord commented:
> if the key is larger than 200, that it does an md5 of the remainder to ensure it stays within the limit.

I recommend using SHA256 as "default" hash instead of MD5. SHA256 is [faster than MD5 since PHP 8.4](https://www.php.net/manual/en/migration84.other-changes.php#migration84.other-changes.performance.hash). SHA256 is also cryptographically secure, unlike MD5.

I am not sure whether collision resistance is needed here for security, but it is generally easier to use SHA256 everwhere than to figure that out.

herbdool commented:
@Sjord I'm just following existing functionality in CiviCRM, including this method https://github.com/civicrm/civicrm-core/blob/85f58c302675fc1a94a6e7de9a38c6e2ad5b2eb0/CRM/Utils/Cache.php#L265 which was just updated a few months ago. I think your recommendation may be beneficial, but that it should be a new issue, so it can be replaced in all locations, rather than hold up this PR.

totten commented:
> I was wondering about the risk with no file locking, or checking a file while it's being updated.

Yeah, tricky question.  We've never had to think about it before because all the existing data-stores for Civi caches (SQL,
Redis, Memcache) have atomic writes as standard-operating-procedure.  But POSIX files require us to be explicit.

I think you've kind of moved past needing an answer, but (for posterity) I suppose the quantiative risk of encountering
corrupt-reads or corrupt-writes varies with factors like:

* The length of time spent processing file-reads/file-writes (aka the duration of the unprotected critical section); eg influenced by:
    * The size of the cache-record, `strlen($value)`. (It takes longer to write 1GB than 1B.) (More bytes => more risk)
    * The type/performance of the underlying filesystem. (Network filesystem, slow filesystem => More risk)
* The frequency of file-updates; eg influenced by
    * The pace of concurrent operations on the system. (More concurrent usage => More risk)
    * The length of the TTL. (Shorter lifetime => More risk)

If we controlled and limited these factors, then one could make the case for calculated-risk.  But realistically, the risks go
up+down depending on independent choices of different actors (e.g.  the site-builder who installs/upgrades an extension; the
constituent who shares a link on a popular website; the staffer who submits a massive dataset; the developer who defines new
caches; the PaaS vendor who changes the filesystem -- all of those people shift the risks without informed-coordination).

> (Re: D8+) ...There are about 3-400 installs for the last four years or so...

> (Re: Laravel) I've been doing some more research.  Turns out that Laravel's file caching includes file locking for
> both reading and writing, similar to D7 version.  So I think I will incorporate it after all.  File caching was the
> default for Laravel until 11.x so they've got lots of experience I think.

Aaah, in terms of practical experience, the Laravel case-study sounds stronger.  :+1:

> * using the LOCK_EX flag when using file_put_contents() - similar to Laravel's approach.
> * added getContents() method which replicates Laravel's and Drupal 7's filecache approach to getting the file content.
> * ...

Sounds pretty good. ~~Jenkins is only complaining about small style issue.~~

totten commented:
> > (@Sjord) SHA256... MD5.... faster.... not sure whether collision resistance is needed here for security...
>
>  (@herbdool) I think your recommendation may be beneficial, but that it should be a new issue, so it can be replaced in all locations, rather than hold up this PR.

Well, faster does sound better.

I agree it looks like a distinct/preexisting issue -- specifically, this formula looks similar to the one in [the Memcached driver](https://github.com/civicrm/civicrm-core/blob/51dda21e6779bb4b8d3e29db638f1974af50915b/CRM/Utils/Cache/Memcached.php#L210-L220). Also worth noting that Laravel's [`FileStore::path($key)`](https://github.com/laravel/framework/blob/v11.42.1/src/Illuminate/Cache/FileStore.php#L347-L352) uses `sha1` (which is also considered an insecure hash nowadays).

So +1 for updating/improving that bit of code. But it's fine as a separate thing.

(Aside: `CRM_Utils_Cache::cleanKey()` is kind of doing a different job -- it accepts old-school keys that were used with `CRM_Core_BAO_Cache::setItem()` and translates those to PSR-16 compliance. But for `Memcached.php` and `FileCache.php`, we expect that the incoming `$key` is already PSR-16 compliant. If a driver does additional filtering, then its purpose is to map from PSR-16-naming to the underlying data-store.)

totten committed:
CRM_Utils_Cache::create() - Allow requesting caches that are specifically file-based

totten committed:
Add E2E_Cache_FileCacheTest

totten commented:
Thanks for keeping this moving, @herbdool!

I did some light `r-run` locally, and it seemed to work for me too. Pushed up a small patch to enable the full tests on this class, but I expect that to work.

MOP
